bpo-37478: Specify possible exceptions for os.chdir()

aeros opened the pull request:
Should this be backported to previous versions of the docs or only applied to 3.9? As far as I'm aware, this exception behavior has occurred since earlier versions of Python 3. These exceptions occur as a result of invalid file paths specified, which was added to ``os.chdir()`` in 3.3, according to the docs.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-37478](https://bugs.python.org/issue37478) -->
https://bugs.python.org/issue37478
<!-- /issue-number -->

aeros committed:
Merge pull request #1 from python/master

Update master

aeros committed:
Add possible exceptions to os.chdir()

aeros committed:
"method" to "function"

aeros committed:
Removed added whitespace

Whitespace added after new section was causing the document tests to fail, this should fix it.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

aeros committed:
Whitespace fix

Removed double space between a period and the next sentence on line 1599.

eryksun commented on Doc/library/os.rst:
```diff
@@ -1598,6 +1598,9 @@ features:
    descriptor must refer to an opened directory, not an open file.
+   
+   This function can raise :exc:`OSError`, such as :exc:`FileNotFoundError` and
+   :exc:`NotADirectoryError`.
```
Another common error is `PermissionError`.  Also, in Windows if a path component exceeds the filesystem's maximum allowed name length or if it contains reserved characters, `chdir` raises `OSError` with `errno.EINVAL`.

It's also noteworthy that `ValueError` is raised if there's a null (`"\x00"`) in the path. However, it might be better to discuss this generally in the section about filenames.

aeros replied:
@eryksun Thanks for the feedback and suggestions. Since ``PermissionError`` falls under the category of ``OSError`` exceptions, I'm not entirely certain as whether or not it should be explicitly mentioned. The purpose of that section was primarily to mention that the function could raise ``OSError`` exceptions and just to provide a couple of examples rather than explicitly state every possible type of ``OSError`` which could be encountered. The section could get excessively verbose rather quickly, see https://docs.python.org/3/library/errno.html#module-errno. 

However, ``PermissionError`` might be common enough to warrant mentioning it, I'll have to see what the others think. As for the ``ValueError`` behavior, since that would apply universally to any file path argument, it should probably be in a more general section (such as the filename section, as you mentioned) rather than in a specific function description.

terryjreedy replied:
I added PermissionError, leaving ValueError for another issue.  Can this function raise OSError itself or only subclasses?  I assumed the later in my edit.

 This function can raise :exc:`OSError` subclasses such as
   :exc:`FileNotFoundError`, :exc:`PermissionError`, and exc:`NotADirectoryError`.

aeros replied:
@terryjreedy Specifically, the function raises the subclasses when it comes to ``FileNotFoundError`` and ``NotADirectoryError``, so your assumption was correct. As far as I am aware, it can't directly raise ``OSError``. I verified this is the case for ``FileNotFoundError`` and ``NotADirectoryError`` in Python 3.9:
![image](https://user-images.githubusercontent.com/44193521/60761479-a2f26f80-a017-11e9-883d-9279b279ce12.png)

aeros replied:
On the commit message for ``Fix exception tag``, it should be ``PermissionError`` instead of ``NotADirectoryError``.

Edit: Not sure if I can retroactively change the commit message, or if that has to be done by a core dev. Let me know if there's a way for me to fix it.

eryksun replied:
> I added PermissionError, leaving ValueError for another issue. Can this function raise OSError itself or only subclasses? I assumed the later in my edit.

As I said in my previous comment, in Windows it can raise `OSError` with `errno == EINVAL` (from WINAPI `winerror == ERROR_INVALID_NAME`) if a path component exceeds the maximum allowed name length or contains characters reserved by the filesystem such as control characters (1-31) and the 5 wildcard characters (`*?"<>`). Python lacks something like an `InvalidArgumentError` subclass for this case. This is a generic filename error like the case with embedded null characters and `ValueError`, so maybe a subsection can briefly discuss invalid filenames in Windows.

terryjreedy replied:
@aeros167 The within PR commit messages do not matter after the fact as commits are squished into 1 for the merge.

@eryksun I commited too soon without fully understanding.  I will do a followup PR adding 'and' to say 'OSError and subclasses'.  This should be good enough for this issue.

terryjreedy committed:
Update os.rst

aeros committed:
Fix exception tag

Where ``NotADirectoryError`` was added to the docs for os.chdir(), there was a missing ``:`` behind ``exc:``

terryjreedy committed:
add space back and trigger retest
